### PR TITLE
Migrate the instructor table to react-table v8

### PR DESCRIPTION
**Pull request by ch-iv**
## Proposed Changes
This PR updates the instructor table from using `react-table` v6 to `@tanstack/react-table` v8. Since the v6 version comes with a UI, but v8 is headless, this introduces the components and styles needed to match the style, feel, and functionality of the old UI.  

image
The table looks the same as before!

## Type of Change
*(Write an `X` or a brief description next to the type or types that best describe your changes.)*

| Type                                                                                    | Applies? |
|-----------------------------------------------------------------------------------------|----------|
| 🚨 *Breaking change* (fix or feature that would cause existing functionality to change) |          |
| ✨ *New feature* (non-breaking change that adds functionality)                          |          |
| 🐛 *Bug fix* (non-breaking change that fixes an issue)                                  |          |
| 🎨 *User interface change* (change to user interface; provide screenshots)              |          |
| ♻️ *Refactoring* (internal change to codebase, without changing functionality)          |          |
| 🚦 *Test update* (change that *only* adds or modifies tests)                            |          |
| 📦 *Dependency update* (change that updates a dependency)                               |          |
| 🔧 *Internal* (change that *only* affects developers or continuous integration)         |     x    |

## Checklist
*(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)*

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] If this is my first contribution, I have added myself to the [list of contributors](https://github.com/MarkUsProject/Markus/blob/master/doc/markus-contributors.txt).

After opening your pull request:

- [x] I have updated the project Changelog (this is required for all changes).
- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments
*(Include any questions or comments you have regarding your changes.)*

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,6 +11,8 @@
     "@rails/ujs": "^7.1.3-4",
     "@rjsf/core": "^5.24.10",
     "@rjsf/validator-ajv8": "^5.24.10",
+    "@tanstack/react-table": "^8.21.3",
+    "react-table": "^6.11.5",
```
Comment (david-yz-liu) [inline package.json]: I don't think you should move this; we want to keep the dependency names in alphabetical order

File: app/javascript/Components/instructor_table.jsx [hunk 2]
```diff
@@ -2,26 +2,58 @@ import React from "react";
...
+        filterFn: "equalsString",
+        meta: {
+          filterVariant: "select",
+        },
+      }),
+      columnHelper.accessor("id", {
+        id: "id",
+        enableSorting: false,
+        header: () => I18n.t("actions"),
+        cell: info => (
```
Comment (david-yz-liu) [inline app/javascript/Components/instructor_table.jsx]: the docs use the parameter name `props`; please use that here to be consistent

File: app/assets/stylesheets/common/_markus.scss [hunk 3]
```diff
@@ -1529,3 +1529,553 @@ canvas {
 .jcrop-centered {
   display: inline-block;
 }
+
+:root {
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_markus.scss]: Let's put this into a separate file and include it here (we do something similar for a few other styled components). There are actually some custom react-table styles we've also included in `_core.scss`; extract them into the new file file too.

Also, please link to a source in the react-table repository.

File: app/javascript/Components/instructor_table.jsx [hunk 4]
```diff
@@ -2,26 +2,58 @@ import React from "react";
...
+        header: () => I18n.t("activerecord.attributes.user.email"),
+      }),
+      columnHelper.accessor("hidden", {
+        header: () => I18n.t("roles.active") + "?",
+        filterFn: "equalsString",
+        meta: {
+          filterVariant: "select",
+        },
+      }),
+      columnHelper.accessor("id", {
```
Comment (david-yz-liu) [inline app/javascript/Components/instructor_table.jsx:38]: This column also shouldn't be searchable

File: app/javascript/Components/table.jsx [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                  </div>
+                );
+              })}
+            </div>
+          ))}
+        </div>
+        <div className="rt-tbody" style={{minWidth: table.getCenterTotalSize()}}>
+          {table.getRowModel().rows.map(row => {
+            return (
+              <div className="rt-tr-group" role="rowgroup">
```
Comment (david-yz-liu) [inline app/javascript/Components/table.jsx]: This needs to have a `key` to prevent the React warning. Is there not a key prop given for the "row group" in the original react table?

File: app/javascript/Components/table.jsx [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                        {flexRender(cell.column.columnDef.cell, cell.getContext())}
+                      </div>
+                    );
+                  })}
+                </div>
+              </div>
+            );
+          })}
+          {!table.getRowModel().rows.length && (
+            <p className="rt-no-data">{noDataText || "No rows found"}</p>
```
Comment (david-yz-liu) [inline app/javascript/Components/table.jsx]: Now that we're integrating the React table code into our codebase, let's create localized versions of the strings (e.g., "No rows found").

File: app/javascript/Components/table.jsx [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    <SelectFilter column={column} />
+  ) : (
+    <SearchFilter column={column} />
+  );
+}
+
+function SearchFilter({column}) {
+  return (
+    <input
+      placeholder={`Search`}
```
Comment (david-yz-liu) [inline app/javascript/Components/table.jsx]: Same comment as above (here and the `aria-label`)

File: app/javascript/Components/table.jsx [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+      type="text"
+      onChange={e => column.setFilterValue(e.target.value)}
+      value={column.getFilterValue()?.toString()}
```
Comment (david-yz-liu) [inline app/javascript/Components/table.jsx]: There's something a bit funny here: when testing this out I see a React warning: 

> Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value change from undefined to a defined value, which should not happen. ...

Maybe use a default value of `""` or something instead of `undefined`?

File: app/javascript/Components/table.jsx [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+    return [...uniqueValuesMap.values()].reduce((sum, count) => sum + count, 0);
+  }, [uniqueValuesMap]);
+
+  return (
+    <select
+      onChange={e => column.setFilterValue(e.target.value)}
+      value={column.getFilterValue()?.toString()}
+      style={{width: "100%"}}
+    >
+      <option value="">All ({totalRowCount})</option>
```
Comment (david-yz-liu) [inline app/javascript/Components/table.jsx]: Same comment as above

File: app/javascript/Components/table.jsx [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+      columnFilters,
+      columnSizing,
+    },
+    onColumnFiltersChange: setColumnFilters,
+    getCoreRowModel: getCoreRowModel(),
+    getFilteredRowModel: getFilteredRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    getFacetedUniqueValues: getFacetedUniqueValues(),
+    getFacetedRowModel: getFacetedRowModel(),
+    debugColumns: false,
```
Comment (david-yz-liu) [inline app/javascript/Components/table.jsx]: is this really necessary (can we leave it out)?

File: app/javascript/Components/table.jsx [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+} from "@tanstack/react-table";
+
+export default function Table({columns, data, noDataText}) {
+  const [columnFilters, setColumnFilters] = React.useState([]);
+  const [columnSizing, setColumnSizing] = React.useState({});
+
+  const table = useReactTable({
+    data,
+    columns,
+    filterFns: {},
```
Comment (david-yz-liu) [inline app/javascript/Components/table.jsx]: is this option necessary?

File: app/javascript/Components/table.jsx [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+                    asc: " rt-resizable-header-content -sort-asc",
+                    desc: " rt-resizable-header-content -sort-desc",
+                  }[header.column.getIsSorted()];
+                return (
+                  <div
+                    className={class_name}
+                    role="columnheader"
+                    tabIndex="-1"
+                    key={header.id}
+                    style={{flex: "100 0 auto", width: header.getSize()}}
```
Comment (david-yz-liu) [inline app/javascript/Components/table.jsx]: move the `flex` styles into the CSS file (I don't see a good reason why they should be inline styles)

File: app/javascript/Components/__tests__/table.test.jsx [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+
+    it("does not display no data message when data exists", () => {
+      render(<Table columns={mockColumns} data={mockData} />);
+
+      expect(screen.queryByText("No rows found")).not.toBeInTheDocument();
+    });
+  });
+
+  describe("Column Sorting", () => {
+    it("sorts column when header is clicked", async () => {
```
Comment (david-yz-liu) [inline app/javascript/Components/__tests__/table.test.jsx]: The current tests here don't actually verify the order of the rows; please add a test for that

File: app/javascript/Components/__tests__/table.test.jsx [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+
+        const values = options.map(option => option.value);
+        const sortedValues = [...values].sort();
+
+        expect(values).toEqual(sortedValues);
+      });
+    });
+  });
+
+  describe("Edge Cases", () => {
```
Comment (david-yz-liu) [inline app/javascript/Components/__tests__/table.test.jsx]: I don't think this should be a separate group; combine with the other `Table` test cases

File: app/javascript/Components/table.jsx [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+          {!table.getRowModel().rows.length && (
+            <p className="rt-no-data">{noDataText || "No rows found"}</p>
+          )}
+        </div>
+      </div>
+    </div>
+  );
+}
+
+function Filter({column}) {
```
Comment (david-yz-liu) [inline app/javascript/Components/table.jsx]: Overall I anticipate the implementation growing as we add more features. This is a great time to create a subfolder and separate each React component into its own file. Eg. `Components/table/table.jsx`, `Components/table/filter.jsx`, etc.

Please also separate the tests into separate files as well

File: package.json (hunk 1 above)
Comment (ch-iv) [reply package.json]: Moved it for a presentation screenshot and forgot to undo. Fixed!

File: app/assets/stylesheets/common/_markus.scss (hunk 3 above)
Comment (ch-iv) [reply app/assets/stylesheets/common/_markus.scss]: Done!

File: app/javascript/Components/table.jsx (hunk 5 above)
Comment (ch-iv) [reply app/javascript/Components/table.jsx]: I don't know if there was a key given for the row group originally. I added `row.id` as the key, which fixed the warning.

File: app/javascript/Components/table.jsx (hunk 10 above)
Comment (ch-iv) [reply app/javascript/Components/table.jsx]: Not at all! We can leave it out.

File: app/javascript/Components/table.jsx (hunk 11 above)
Comment (ch-iv) [reply app/javascript/Components/table.jsx]: This option is not necessary.

File: app/javascript/Components/table.jsx (hunk 12 above)
Comment (ch-iv) [reply app/javascript/Components/table.jsx]: Done!

File: app/javascript/Components/__tests__/table.test.jsx [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+      expect(rows[2]).toHaveTextContent("Bob Johnson");
+      expect(rows[3]).toHaveTextContent("Alice Brown");
+    });
+  });
+
+  describe("No Data Handling", () => {
+    it("displays default no data message when data is empty", () => {
+      render(<Table columns={mockColumns} data={[]} />);
+
+      expect(screen.getByText("No rows found")).toBeInTheDocument();
```
Comment (david-yz-liu) [inline app/javascript/Components/__tests__/table.test.jsx]: Internationalize the "No rows found" in this test file as well

File: app/javascript/Components/__tests__/table.test.jsx [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+      expect(selectFilter.value).toBe("active");
+    });
+  });
+
+  describe("Filter Components", () => {
+    describe("SearchFilter", () => {
+      it("renders with correct attributes", () => {
+        render(<Table columns={mockColumns} data={mockData} />);
+
+        const searchInputs = screen.getAllByPlaceholderText("Search");
```
Comment (david-yz-liu) [inline app/javascript/Components/__tests__/table.test.jsx]: As above, make sure "Search" is now internationalized as well

File: app/javascript/Components/__tests__/table.test.jsx [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+      expect(nameHeader.closest(".rt-th")).toHaveClass("rt-resizable-header-content");
+    });
+
+    it("applies sorting classes to all headers", () => {
+      render(<Table columns={mockColumns} data={mockData} />);
+
+      const emailHeader = screen.getByText("Email");
+      const headerElement = emailHeader.closest(".rt-th");
+
+      expect(headerElement).toHaveClass("--cursor-pointerundefined");
```
Comment (david-yz-liu) [inline app/javascript/Components/__tests__/table.test.jsx]: the `undefined` is quite janky; please modify the implementation so that no class is added if the sort order isn't `asc`/`desc`

**Review by david-yz-liu — COMMENTED** [review]
@ch-iv looks great! I just left a few minor comments.

**Comment by ch-iv** [comment]
I realized that I vibed a little too much when writing the tests. I've rewritten them to be decent. 

Also, sorry for the messy git history -- I had to `npm install` to refresh the package lock mid rebase.

**Review by david-yz-liu — APPROVED** [review]
Great work, @ch-iv!
